pageserver/src/tenant/timeline/detach_ancestor.rs

```diff
@@ -116,6 +130,472 @@ impl Default for Options {
     }
 }
 
+/// SharedState manages the pausing of background tasks (GC) for the duration of timeline detach
+/// ancestor.
+///
+/// Currently this is tracked at tenant level, but it could be moved to be on the roots
+/// of each timeline tree.
+#[derive(Default)]
+pub(crate) struct SharedState {
+    inner: std::sync::Mutex<SharedStateInner>,
+    gc_waiting: tokio::sync::Notify,
+    attempt_waiting: tokio::sync::Notify,
+}
+
+impl SharedState {
+    /// Notify an uninitialized shared state that an attempt to detach timeline ancestor continues
+    /// from previous instance.
+    pub(crate) fn continue_existing_attempt(&self, attempt: &Attempt) {
```

Comment:
During normal, uninterrupted detach, this method is called on state initialized by `start_new_attempt`.

Unless you and I have a different definition of "initialized"

Reply:
> During normal, uninterrupted detach

We have two tenants in play here:
- tenant before `reset_tenant`, let's call it "tenant_1"
- tenant after `reset_tenant`, let's call it "tenant_2"

1. start_new_attempt on "tenant_1"
2. (prepare completes)
3. (detach_and_reparent completes)
4. reset_tenant completes
    - now we have shutdown the "tenant_1"
    - we've created the "tenant_2" in `crate::tenant::mgr::tenant_spawn`
    - this method is called

I am calling the SharedState uninitialized because what will happen is:

1. SharedState::continue_existing_attempt (before the actual task to load and activate is spawned)
2. SharedStateBuilder visits all timelines index_part.json
3. SharedStateBuilder completes (and "initializes" the SharedState)

Reply:
This is now inlined in before upserting the "tenant_2" into the tenant map, very much likely always happening before attach gets to start (but not guaranteed). It doesn't matter, because we re-establish the "attempt exclusivity" there.

Not linking any commit now because I might be able to split easy PRs before you get to re-review this.